[Sema] Implement basic type checking for pack expansion expressions.

hborla opened the pull request:
This change rewrites postfix `...` operator expressions to `PackExpansionExpr` when the operand contains pack references. Pack references are replaced with `OpaqueValueExpr`s, and the `PackExpansionExpr` stores the bindings from opaque values to pack references.

This is currently done while pre-checking an expression before constraint generation, but this could later be moved into the solver by rewriting the postfix operator expression to an `UnresolvedExpansionOperator` or similar that generates a disjunction between the between postfix `...` and expansion operator, and delaying constraint generation for the operand until after a disjunction choice is attempted.

hborla committed:
[AST] Store opaque value bindings in PackExpansionExpr.

hborla committed:
[ConstraintSystem] Implement basic type checking for pack expressions.

During prechecking, postfix '...' expressions are rewritten to pack
expansion expressions if the operand contains pack references. References
to packs are replaced with opaque values, and the pack expansion stores
the opaque value bindings to decl refs.

hborla committed:
[ConstraintSystem] Implement value parameter pack forwarding.

slavapestov commented:
Heads up, I merged https://github.com/apple/swift/pull/61657 so you can drop your cherry pick

hborla committed:
[AST] Plumb reduced shape types through local requirements into pack archetypes.

hborla committed:
[RequirementMachine] Use pack expansion types as a source for requirement inference.

Previously, only the pattern type was added as a requirement inference source, but
the pack expansion type is necessary for inferring same-shape requirements for
packs that are expanded in parallel.

hborla committed:
[ConstraintSystem] Use reduced shape types when generating constraints for pack
expansion expressions.

hborla committed:
[CSApply] Implement coerceToType for pack expansion expressions.

hborla committed:
[AST] Allow tuple types with consecutive pack expansions.

hborla committed:
[GenericEnvironment] Map shape types into context in PackArchetypeType::getShape.

slavapestov commented on lib/AST/GenericEnvironment.cpp:
```diff
@@ -365,6 +365,7 @@ GenericEnvironment::getOrCreateArchetypeFromInterfaceType(Type depType) {
   if (rootGP->isParameterPack()) {
     assert(getKind() == Kind::Primary);
     result = PackArchetypeType::get(ctx, this, requirements.anchor,
+                                    requirements.packShape,
```
Should we map the packShape into the generic environment to make it distinct for distinct archetypes? (We also map the superclass type into context)

slavapestov commented on test/Generics/pack-shape-requirements.swift:
```diff
@@ -59,3 +59,9 @@ struct Ts<T...> {
+
+// CHECK-LABEL: expandedParameters(_:transform:)
+// CHECK-NEXT: Generic signature: <T..., Result... where T.shape == Result.shape>
+func expandedParameters<T..., Result...>(_ t: T..., transform: ((T) -> Result)...) -> (Result...) {
```
Can you add a third generic parameter and change the outer return type, to ensure that we don't equate the shapes of all three generic parameters?

hborla commented on lib/Sema/PreCheckExpr.cpp:
```diff
@@ -414,6 +414,50 @@ static bool exprLooksLikeAType(Expr *expr) {
+          auto sourceRange = declRef->getSourceRange();
+
+          // FIXME: The OpaqueValueExpr should use the opened element
+          // type of the pattern.
```
I'm still working on this, but I'll do it in a follow-up PR

hborla replied:
Nevermind!

hborla commented:
@swift-ci please smoke test

hborla committed:
[GenericSignature] Correct the order of arguments to GenericTypeParamType::get
when creating the generic signature for an opened pack element.

hborla committed:
[ConstraintSystem] Use opened element types for opaque values in pack expansions.

hborla committed:
[ConstraintSystem] Add a PackElementOf constraint to delay mapping opened
element pattern types to pattern archetypes until after type variables are
resolved.

hborla commented:
@swift-ci please smoke test

hborla committed:
[PreCheckExpr] Gate PackExpansionExpr creation behind the variadic generics flag.

hborla commented:
@swift-ci please smoke test

hborla commented:
@swift-ci please smoke test macOS

xedin commented on lib/Sema/CSGen.cpp:
```diff
@@ -2874,7 +2881,31 @@ namespace {
     Type visitPackExpansionExpr(PackExpansionExpr *expr) {
-      llvm_unreachable("not implemented for PackExpansionExpr");
+      for (auto *binding : expr->getBindings()) {
+        auto type = visit(binding);
```
This might be an oversight - `Traversal::visitPackExpansionExpr` in ASTWalker.cpp needs to be updated to walk into `bindings` and re-set them, same applies to ASTDumper and ASTPrinter most likely.

hborla replied:
No, not visiting these in ASTWalker was intentional. This follows the same pattern as `OpenExistentialExpr` which only visits the subexpression containing opaque values, and the bindings are handled separately where needed. These aren't actually separate values in the AST, but we do need the types in the constraint system.

xedin replied:
Ok, we probably need a comment about this then... ASTDumper still needs to be updated to print bindings and opaque values just like `OpenExistentialExpr` does.

xedin commented on lib/Sema/CSApply.cpp:
```diff
@@ -3802,7 +3818,13 @@ namespace {
-      llvm_unreachable("not implemented for PackExpansionExpr");
+      for (unsigned i = 0; i < expr->getNumBindings(); ++i) {
+        auto *binding = expr->getBindings()[i];
+        expr->setBinding(i, visit(binding));
```
Can `visit` add any implicit AST nodes for binding? I think it might be safer to add `cs.cacheTypes(<result>)` in this case.

xedin commented on lib/Sema/CSApply.cpp:
```diff
@@ -3598,7 +3598,23 @@ namespace {
+
+      // A ParenExpr can end up with a tuple type if it contains
+      // a pack expansion. Rewrite it to a TupleExpr.
+      if (dyn_cast<PackExpansionExpr>(expr->getSubExpr())) {
```
Nit: `isa`

hborla replied:
Thanks! I forgot about `isExpr` 🙂

xedin commented on lib/Sema/CSApply.cpp:
```diff
@@ -3598,7 +3598,23 @@ namespace {
+                                   /*elementNames=*/{},
+                                   /*elementNameLocs=*/{},
+                                   expr->getRParenLoc(),
+                                   expr->isImplicit());
```
I think this needs `cacheTypes` as well.

xedin replied:
Actually never mind, `setType` is good enough.

xedin commented on lib/Sema/CSSimplify.cpp:
```diff
@@ -8391,6 +8428,43 @@ ConstraintSystem::simplifyBindTupleOfFunctionParamsConstraint(
+                                                  TypeMatchOptions flags,
+                                                  ConstraintLocatorBuilder locator) {
+  auto elementType = simplifyType(first, flags);
+  auto *loc = getConstraintLocator(locator);
```
Nit: Might be better to materialize locator only when we are avoid to add unsolved constraint since the other case takes a builder anyway...

xedin commented on lib/Sema/CSApply.cpp:
```diff
@@ -7066,11 +7087,54 @@ Expr *ExprRewriter::coerceToType(Expr *expr, Type toType,
+    auto shapeType = toExpansionType->getCountType();
+    auto expansionTy = PackExpansionType::get(patternType, shapeType);
+
+    return cs.cacheType(PackExpansionExpr::create(ctx, pattern,
```
Instead of creating new expansion expression it could be updated with new type and pattern in-place?

hborla replied:
It could, sure, but I'm working on moving `...` resolution into the solver, so `PackExpansionExpr`s will only be created in CSApply anyway

xedin replied:
Sounds good then!
